From 44826337cc5006b5f9916761c52aa937e51cd4f0 Mon Sep 17 00:00:00 2001
From: Rander Wang <rander.wang@linux.intel.com>
Date: Mon, 19 Mar 2018 13:39:49 +0800
Subject: [PATCH 0867/1240] ASoC: Intel: fix playback only once bug

Host cant get interrupt when playback after first time
because the ack register for interrupt is set two times
in driver, one in IRQ function and the other is in cmd
done function which is scheduled by IRQ function.This
make interrupt in dsp HW not work. Now accept interrupt
in IRQ function and acknowledge to dsp when cmd done

Signed-off-by: Rander Wang <rander.wang@linux.intel.com>
---
 sound/soc/sof/hw-apl.c | 17 +++--------------
 1 file changed, 3 insertions(+), 14 deletions(-)

diff --git a/sound/soc/sof/hw-apl.c b/sound/soc/sof/hw-apl.c
index 3db9e6c..ca84b5e 100644
--- a/sound/soc/sof/hw-apl.c
+++ b/sound/soc/sof/hw-apl.c
@@ -1391,18 +1391,13 @@ static irqreturn_t cnl_irq_thread(int irq, void *context)
 		/* handle messages from DSP */
 		snd_sof_ipc_msgs_rx(sdev);
 
-		/* clear busy interrupt */
+		/* clear busy interrupt to tell dsp controller this */
+		/* interrupt has been accepted, not trigger it again */
 		snd_sof_dsp_update_bits_forced(sdev, APL_DSP_BAR,
 					       CNL_DSP_REG_HIPCTDR,
 					       CNL_DSP_REG_HIPCTDR_BUSY,
 					       CNL_DSP_REG_HIPCTDR_BUSY);
 
-		/* set done bit to ack dsp */
-		snd_sof_dsp_update_bits_forced(sdev, APL_DSP_BAR,
-					       CNL_DSP_REG_HIPCTDA,
-					       CNL_DSP_REG_HIPCTDA_DONE,
-					       CNL_DSP_REG_HIPCTDA_DONE);
-
 		ret = IRQ_HANDLED;
 	}
 
@@ -1438,13 +1433,7 @@ static int apl_cmd_done(struct snd_sof_dev *sdev)
 
 static int cnl_cmd_done(struct snd_sof_dev *sdev)
 {
-	/* clear busy interrupt */
-	snd_sof_dsp_update_bits_forced(sdev, APL_DSP_BAR,
-				       CNL_DSP_REG_HIPCTDR,
-				       CNL_DSP_REG_HIPCTDR_BUSY,
-				       CNL_DSP_REG_HIPCTDR_BUSY);
-
-	/* set done bit to ack dsp */
+	/* set done bit to ack dsp the msg has been processed */
 	snd_sof_dsp_update_bits_forced(sdev, APL_DSP_BAR,
 				       CNL_DSP_REG_HIPCTDA,
 				       CNL_DSP_REG_HIPCTDA_DONE,
-- 
2.7.4

